File: common/src/main/java/com/hartwig/actin/molecular/interpretation/AggregatedEvidence.java
package com.hartwig.actin.molecular.interpretation;

import com.google.common.collect.Multimap;
import com.hartwig.actin.molecular.datamodel.evidence.ExternalTrial;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class AggregatedEvidence {

    @NotNull
    public abstract Multimap<String, String> approvedTreatmentsPerEvent();

    @NotNull
    public abstract Multimap<String, ExternalTrial> externalEligibleTrialsPerEvent();

    @NotNull
    public abstract Multimap<String, String> onLabelExperimentalTreatmentsPerEvent();

    @NotNull
    public abstract Multimap<String, String> offLabelExperimentalTreatmentsPerEvent();

    @NotNull
    public abstract Multimap<String, String> preClinicalTreatmentsPerEvent();

    @NotNull
    public abstract Multimap<String, String> knownResistantTreatmentsPerEvent();

    @NotNull
    public abstract Multimap<String, String> suspectResistanceTreatmentsPerEvent();

}

File: common/src/main/kotlin/com/hartwig/actin/molecular/interpretation/AggregatedEvidence.kt
package com.hartwig.actin.molecular.interpretation

import com.hartwig.actin.datamodel.molecular.evidence.ExternalTrial
import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidence
import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidenceCategories.approved
import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidenceCategories.experimental
import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidenceCategories.knownResistant
import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidenceCategories.preclinical
import com.hartwig.actin.datamodel.molecular.evidence.TreatmentEvidenceCategories.suspectResistant

data class AggregatedEvidence(
    val treatmentEvidencePerEvent: Map<String, Set<TreatmentEvidence>> = emptyMap(),
    val eligibleTrialsPerEvent: Map<String, Set<ExternalTrial>> = emptyMap(),
) {
    fun approvedTreatmentsPerEvent() = filterMap { approved(it.value) }
    fun onLabelExperimentalTreatmentPerEvent() = filterMap { experimental(it.value, true) }
    fun offLabelExperimentalTreatmentsPerEvent() = filterMap { experimental(it.value, false) }
    fun preClinicalTreatmentsPerEvent() = filterMap { preclinical(it.value, true) }
    fun knownResistantTreatmentsPerEvent() = filterMap { knownResistant(it.value, true) }
    fun suspectResistantTreatmentsPerEvent() = filterMap { suspectResistant(it.value, true) }

    private fun filterMap(mappingFunction: (Map.Entry<String, Set<TreatmentEvidence>>) -> List<TreatmentEvidence>) =
        treatmentEvidencePerEvent.mapValues(mappingFunction).filterValues { it.isNotEmpty() }
}
